### ISSUE TITLE
LightGBMClassifier, LightGBMRegressor hang indefinitely without error at fit()

### ISSUE THREAD

**Issue by mcb0035** [Azure/mmlspark#623]
LightGBMClassifier and LightGBMRegressor both hang indefinitely most of the time on Databricks.  My 6 node cluster has Databricks runtime 5.5, Scala 2.11, Spark 2.4.3, and mmlspark_2.11:0.17.dev27.  I started using the development version of mmlspark after seeing comments claiming several issues in 0.17 were fixed there, but the problem persists.  Another comment suggested adding useBarrierExecutionMode=True, so I did that as well, but that also did not fix the problem.

Usually there is no error, it just hangs on fit() for hours or days until I kill the job, with nothing output to stdout or stderr.  Occasionally it fails with "connection refused" error, but as stated elsewhere, this is a generic error that does not point to the root cause.

A few times the classifier successfully trained a model, which took about 1 second per iteration for the same inputs.  The inconsistency, lack of error messages, and occasional success for the same inputs make this difficult to debug.  I will update if I get a more informative error message.

**Comment by imatiach-msft** [comment]
@mcb0035 Did you disable dynamic allocation?  We don't support dynamic allocation yet.
For classification case, might you have unbalanced data?  if so, could you try out this build, where I throw an error on unbalanced classes, which comes from this PR, https://github.com/Azure/mmlspark/pull/618:
--packages
com.microsoft.ml.spark:mmlspark_2.11:0.17+83-11237da2
and --repositories
https://mmlspark.azureedge.net/maven

for LightGBMRegressor, I'm not sure why you are seeing issues.  What parameters are you running on databricks with?

**Comment by imatiach-msft** [comment]
![image](https://user-images.githubusercontent.com/24683184/61567197-bad0e700-aa4c-11e9-975c-58602d9ce5ca.png)

**Comment by mcb0035** [comment]
Autoscaling is turned off for this cluster.  I will try the new build and report back.

**Comment by mcb0035** [comment]
Also, [documentation for Azure Machine Learning](https://docs.microsoft.com/en-us/azure/machine-learning/service/how-to-configure-environment#aml-databricks), which we intend to drive this workflow, says to only use non-ML Databricks runtime.  Are there any additional libraries that need to be imported for mmlspark to work with the non-ML runtime?

**Comment by imatiach-msft** [comment]
hi @mcb0035 , sorry about the trouble you are having, the imports have changed after the refactor by @mhamilton723 , the new imports are:
```
from mmlspark.train import LightGBMClassifier, LightGBMRegressor
```

**Comment by mcb0035** [comment]
That did not work either:
![image](https://user-images.githubusercontent.com/8672657/61665826-e2ad8e00-ac8a-11e9-8de3-4229cf577a8a.png)

This is for mmlspark_2.11:0.17+83-11237da2.

**Comment by imatiach-msft** [comment]
sorry, it's from mmlspark.lightgbm import LightGBMClassifier

**Comment by imatiach-msft** [comment]
@mcb0035 did you run into any other issues?  would you be able to close this issue?  It sounds like we were able to resolve it when debugging by not using the new barrier execution mode (which doesn't seem to be working).

**Comment by mcb0035** [comment]
Yes, useBarrierExecutionMode=False and isUnbalance=True solved the problem.  Thanks so much for your help.

**Comment by imatiach-msft** [comment]
@mcb0035 can this issue be closed or are there any other problems you are still encountering?

**Comment by anttisaukko** [comment]
With the barrierExecutionModel=true, LightGBMClassifier() will randomly hang without an error (tested with 671b68892ace5967e60c7a064effd42dd5a21ec7).

**Comment by mholmeslinder** [comment]
I will also post an issue with more details for this, but I'm having the same issue with the SynapseML Vowpal Wabbit `VowpalWabbitContextualBandit()` model. Setting `useBarrierExecutionMode=False` and turning Auto-Scaling off didn't resolve it.

This ```Usually there is no error, it just hangs on fit() for hours or days until I kill the job, with nothing output to stdout or stderr.

A few times the classifier successfully trained a model, which took about 1 second per iteration for the same inputs. The inconsistency, lack of error messages, and occasional success for the same inputs make this difficult to debug. I will update if I get a more informative error message``` describes the behavior perfectly.

Here is the basic cluster config (with autoscaling turned back on, since turning it off didn't help).

<img width="597" alt="image" src="https://user-images.githubusercontent.com/12013515/233446419-07a54448-84e1-434c-b3e6-6c0b2bf86ea6.png">

**Comment by ranadeepsingh** [comment]
A network thread left blocked in `accept()` with no error raised is precisely the `fit()`-hangs-forever signature reported here, so flagging it.

PR #2612 (`fix: preserve the real cause when a LightGBM task retry cannot rejoin the network`) just merged, and it fixes two concrete socket-lifecycle defects that can produce this signature. **I am deliberately not closing this issue** — I cannot prove these were your root cause — but they are worth retesting against.

What actually changed in `NetworkManager` / `LightGBMBase`:

1. **The driver network thread could stay blocked in `accept()` holding the driver port.** If the partition tasks failed, `executeTraining` never closed the network manager, so the thread sat waiting for connections that would never arrive, for up to the configured timeout. It is now closed in a `finally`.
2. **The driver server socket was leaked** when the topology round threw or timed out. It is now closed in a `finally` too.
3. **`closeConnections` is now idempotent and synchronized**, closes the load-only helper sockets that were previously leaked on *every* `fit`, and closes each socket independently so one failure cannot strand the rest.
4. On a terminal `ConnectException`, the error now explains that the topology exchange is already closed and that the real cause is in the logs of the **first** failed attempt of that partition. Previously Spark only surfaced the last attempt, which was always a bare `Connection refused` — masking the original failure.

Item 4 is the one most likely to help you make progress: if this still reproduces, the next release should now tell you *why* the first attempt died instead of hiding it.

If anyone here can retest once this ships and report back, that would give us what we need to either close this or finally pin down the remaining cause.

**Comment by ranadeepsingh** [comment]
Thank you for the detailed history here. PR #2612 fixed two concrete LightGBM socket-lifecycle defects that can cause this signature, but it does not prove every hang is resolved. Keeping this open and asking for a retest on v1.1.3 rather than closing it as a duplicate of #1967.
